Fix nullable deleted column in API Keys table

davelopez opened the pull request:
By default, the `deleted` column is nullable even if it has a [default value of False](https://github.com/galaxyproject/galaxy/blob/1a984f27ab755b71a817f2bb95139af9fd43c7aa/lib/galaxy/model/__init__.py#L9700).

There is a chance, some external scripts that create API keys for users directly in the database might not set the `deleted` value explicitly causing possible incongruences with multiple API keys for the user not being explicitly deleted=False.

This restricts those cases by forcing a non-null value.

We could also try to handle possible null values in the API logic, I can change the approach if this is more desirable than altering the database schema at this point.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek commented on lib/galaxy/model/migrations/alembic/versions_gxy/1d0564b6d7aa_make_api_keys_deleted_non_nullable.py:
```diff
@@ -0,0 +1,32 @@
+def upgrade():
+    # Update any existing rows with a deleted=NULL value to be False.
+    table = sa.sql.table(table_name, sa.Column(column_name, sa.Boolean(), nullable=True))
+    op.execute(table.update().where(table.c.deleted.is_(None)).values(deleted=False))
```
If you do want to alter these you'll have to make it so that only the newest API key is set to `deleted=False` and all older ones are set to `deleted=True`.

mvdbeek replied:
and of course only if there isn't one that is already valid for that user.

hexylena replied:
```suggestion
    op.execute(table.update().where(table.c.deleted.is_(None)).values(deleted=True))
```

even without the sorting, safer to just have it fail closed.

davelopez replied:
Done in the rebase to keep the migration script in a single commit, thanks!

davelopez committed:
Set APIKeys.deleted non-nullable in model

mvdbeek reviewed (approved):
Looks good to me, @jdavcs can you also take a look ?

davelopez committed:
Add db migration script APIKeys non-null deleted

This makes sure the deleted column in the `api_keys` table has a valid boolean value from now on and sets any existing null value as deleted (True).

nsoranzo commented on lib/galaxy/model/migrations/alembic/versions_gxy/b855b714e8b8_make_api_keys_deleted_non_nullable.py:
```diff
@@ -0,0 +1,39 @@
+
+
+def downgrade():
+    alter_column(table_name, column_name, existing_type=sa.Boolean(), nullable=True, server_default=None)
```
```suggestion
    alter_column(table_name, column_name, existing_type=sa.Boolean(), nullable=True, existing_server_default=False)
```

davelopez replied:
This will modify the original column's behavior after a downgrade, right?
With this proposed change, after the `downgrade` any new "deleted=null" values will be defaulted to "deleted=False". I don't think this is what we want unless I'm missing something :thinking:

nsoranzo replied:
`False` is the current and future default for the `deleted` column.

nsoranzo replied:
I was mistaken here, we set a `default` but not a `server_default` for this column (actually for any column). Thanks to David, I discovered `server_default` is the one that should be used if you want the default to apply also outside of SQLAlchemy (i.e. when using SQL statements).

jdavcs replied:
`False` is the default in Alembic's `alter_column` function. However, in SQLAlchemy, the default is `None` ([Column costructor](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Column.__init__)). Thus, if we run this migration, Alembic's default is applied, but if we skip migrations (e.g. new installation), it'll be SQLAlchemy's default. Therefore, given that we want the downgrade function to reverse the upgrade, we want the database to have `null` as the default value, which (in the case of postgres) is done by dropping the default (not setting it to false). At least that's my understanding.

jdavcs replied:
More on this :point_up: Essentially, even though the resulted value in the database will be the same, `default` is not the same as `server_default` (I'm looking at [these docs](https://docs.sqlalchemy.org/en/20/core/defaults.html)). Since we do not use `server_default` in the model's definition (and why would we), I think it would make sense to ensure the column DDL in the database does not change as a result of running an upgrade->downgrade sequence?

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -10140,7 +10140,7 @@ class APIKeys(Base, RepresentById):
     key = Column(TrimmedString(32), index=True, unique=True)
     user = relationship("User", back_populates="api_keys")
-    deleted = Column(Boolean, index=True, default=False)
+    deleted = Column(Boolean, index=True, default=False, nullable=False)
```
```suggestion
    deleted = Column(Boolean, index=True, server_default=False, nullable=False)
```

davelopez committed:
Set server_default in the model definition

This makes sure the default value is set when inserting new rows with or without using Sqlalchemy.

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>
